lib/common/util/exceptions.js

```diff
@@ -0,0 +1,69 @@
+// @flow
+
+import moment from 'moment'
+
+import type {ExceptionDate, ScheduleExceptionDateRange} from '../../types'
+
+/**
+ * This method updates the current dates based on the updated range. The method determines if the range has been extended or
+ * condensed and adds or subtracts dates as appropriate.
+ */
+export function updateDates (isStartDate: boolean, newRangeBoundary: moment, dates: Array<string>, startMoment: moment, endMoment: moment) {
+  if (isStartDate) {
+    const rangeExtended = startMoment.diff(newRangeBoundary) > 0
+    return rangeExtended
+      ? modifyRangeOfDates(dates, endMoment, newRangeBoundary, false)
+      : modifyRangeOfDates(dates, newRangeBoundary.subtract(1, 'days'), startMoment, true) // Keep the end date
+  } else {
+    const rangeExtended = newRangeBoundary.diff(endMoment) > 0
+    return rangeExtended
+      ? modifyRangeOfDates(dates, newRangeBoundary, startMoment, false)
+      : modifyRangeOfDates(dates, endMoment, newRangeBoundary.add(1, 'days'), true) // Keep the start date
+  }
+}
+
+/**
+ * This method adds or removes a range of dates from startRange to endRange to the dates array.
+ */
+export function modifyRangeOfDates (
+  dates: Array<string>,
+  endRange: moment,
+  startRange: moment,
+  remove: boolean
+) {
+  const end = endRange.format('YYYYMMDD')
+  const start = startRange.format('YYYYMMDD')
+
+  // $FlowFixMe: Flow doesn't understand sets
+  if (startRange.unix() > endRange.unix()) return Array.from(new Set(dates))
+
+  if (remove) {
+    dates = dates.filter(el => el !== end && el !== start)
+  } else {
+    dates.push(end, start)
+  }
+
+  endRange.subtract(1, 'days')
+  startRange.add(1, 'days')
+
+  return modifyRangeOfDates(dates, endRange, startRange, remove)
+}
+
+/**
+ * This method parses dates to extract ranges and individual exception dates. It assumes that dates is sorted.
+ */
+export function getRangesForDates (dates: Array<ExceptionDate>, ranges: Array<ScheduleExceptionDateRange> = []) {
+  // If dates is empty, return ranges
+  if (dates.length === 0) return ranges
+  // Grab first date, iterate over ranges and see if any startRange is above the date
+  const date = dates.shift()
+  if (date === 0) ranges.push({startDate: 0, endDate: 0})
```

Comment:
This method is well commented so we should probably add one more comment here that says why we did this

Reply:
updated in https://github.com/ibi-group/datatools-ui/pull/828/commits/b8ea974dc8ba296b92d2b964760b893be231af36